monkey/monkey_island/cc/services/agent_plugin_service/mongo_agent_plugin_repository.py

```diff
@@ -0,0 +1,226 @@
+        self, mongo_client: MongoClient
+    ) -> Dict[OperatingSystem, gridfs.GridFS]:
+        agent_plugins_binaries_collections: Dict[OperatingSystem, gridfs.GridFS] = {}
+
+        for os in OperatingSystem:
+            agent_plugins_binaries_collections[os] = gridfs.GridFS(
+                mongo_client.monkey_island, self._get_binary_collection_name(os)
+            )
+
+        return agent_plugins_binaries_collections
+
+    def _get_binary_collection_name(self, operating_system: OperatingSystem) -> str:
+        return f"agent_plugins_binaries_{operating_system.value}"
+
+    def get_plugin(
+        self, host_operating_system: OperatingSystem, plugin_type: AgentPluginType, name: str
+    ) -> AgentPlugin:
+        try:
+            plugin_dict = self._get_agent_plugin(plugin_type, name)
+        except Exception:
+            raise UnknownRecordError(
+                f"Error retrieving the agent plugin {name} of type {plugin_type}"
+            )
+
+        try:
+            os_binaries = plugin_dict[BINARY_OS_MAPPING_KEY]
+            _id = os_binaries[host_operating_system.value]
+            plugin_source_bytes = (
+                self._agent_plugins_binaries_collections[host_operating_system].get(_id).read()
+            )
+
+            plugin_dict.pop(BINARY_OS_MAPPING_KEY)
+            plugin_dict.pop("_id")
+            plugin_dict["source_archive"] = plugin_source_bytes
+            return AgentPlugin(**plugin_dict)
+        except Exception as err:
+            raise RetrievalError(
+                f"Error retrieving the agent plugin {name} of type {plugin_type} for operating "
+                f"system {host_operating_system}: {err}"
+            )
+
+    def _get_agent_plugin(self, plugin_type: AgentPluginType, plugin_name: str) -> Dict[str, Any]:
+        plugin_dict = self._agent_plugins_collection.find_one(
+            {"plugin_manifest.name": plugin_name, "plugin_manifest.plugin_type": plugin_type.value}
+        )
+        if plugin_dict is None:
+            raise RuntimeError("Not found")
+        return plugin_dict
+
+    def get_all_plugin_configuration_schemas(
+        self,
+    ) -> Dict[AgentPluginType, Dict[str, Dict[str, Any]]]:
+        configuration_schema_dicts = self._agent_plugins_collection.find(
+            {}, {"plugin_manifest.plugin_type": 1, "plugin_manifest.name": 1, "config_schema": 1}
+        )
+        configuration_schemas: Dict[AgentPluginType, Dict[str, Dict[str, Any]]] = {}
+
+        for item in configuration_schema_dicts:
+            try:
+                plugin_type = AgentPluginType(item["plugin_manifest"]["plugin_type"])
+            except ValueError:
+                raise RetrievalError(
+                    f"Invalid plugin type stored in the database:"
+                    f" {item['plugin_manifest']['plugin_type']}"
+                )
+            plugin_name = item["plugin_manifest"]["name"]
+            config_schema_dict = item["config_schema"]
```

Comment:
`KeyErrors` need to be caught and wrapped as `RetrievalErrors`.

Reply:
If these keys don't exists the schema won't be pulled in the first place